Add explicit interface for saving history with or without updating current session (backport)

azat opened the pull request:
Before this patch replxx, while saving to the history file always reload
the history from the disk and add new items from that file into the
current session. This looks incovenient, if user code calls
history_save() after each command (to save history ASAP, example of such
program is clickhouse-client) since if you run two programs with replxx
completion the completion will overlaps after each command.

Looks like at least this should be configurable.

Refs: https://github.com/ClickHouse/ClickHouse/pull/13086
Upstream PR: https://github.com/AmokHuginnsson/replxx/pull/83

alexey-milovidov commented:
But actually it is very convenient.

In other programs that are using readline

It was in the list of demanded features:

alexey-milovidov commented:
But actually it is very convenient.
In other programs that are using readline two programs run in parallel will rewrite history of each other (data loss).

It was in the list of demanded features:
https://clickhouse.tech/docs/ru/whats-new/extended-roadmap/#uluchshenie-vozmozhnostei-interaktivnogo-rezhima-clickhouse-client

alexey-milovidov commented:
If history update is implemented as append under flock, it's Ok.

azat commented:
>In other programs that are using readline two programs run in parallel will rewrite history of each other (data loss).

You are right, it should do reload from file, but do not add new items from file to the completion list in the current client.
At least this is what I'm interested in, thoughts?

alexey-milovidov commented:
> You are right, it should do reload from file, but do not add new items from file to the completion list in the current client.

Why need to reload? Just append to history file...

azat commented:
>Why need to reload? Just append to history file...

But in this case you may have duplicates in file (and besides it requires tracking of which was written, or saving some offset)

azat commented:
Make it always reads from disk, w/o updating current client history (this PR and upstream)

azat commented:
Plus, adds `.hist_autosave` command for `replxx-example-cxx-api` that behaves as clickhouse-client, i.e. calls `history_save` after each command

azat commented:
@alexey-milovidov thoughts on this?

AmokHuginnsson committed:
Add explicit interface for saving history with or without updating
current session.

Before this patch replxx, while saving to the history file always reload
the history from the disk and add new items from that file into the
current session. This could be incovenient, if user code calls
history_save() after each command (to save history ASAP, example of such
program is clickhouse-client) since if you run two programs with replxx
completion the completion will overlaps after each command.

Hence an introduction of additional interface:
Replxx::history_save() - just save merged history without changing
current session
Replxx::history_sync() - save merged history and update current session

Upstream commit is abe46eabaa7320a5fcb72788e74a87339884b408 ("Add
explicit interface for saving history with or without updating current
session.")

And this also partially backports
737c8e8147d44eae530e2b56592ad8788695ef12 ("History load/save methods
return success status.")

azat commented:
This has been implemented in upstream (in a little bit different way - abe46eabaa7320a5fcb72788e74a87339884b408), so after it history_save() will only write to the file, w/o updating current session, but that patch requires backporting since this fork has some modifications, I will replace this patch with backporting upstream commit

alexey-milovidov commented:
Ok.

> @alexey-milovidov thoughts on this?

No thoughts, I forgot about this PR.

alexey-milovidov commented:
I don't remember, why it's not good to have single persistent shared history?
So, when you work with many clickhouse-clients, they will see history of each other.
It's very convenient when you just switch to some random tab and want to get a query from the history.

azat commented:
> I don't remember, why it's not good to have single persistent shared history?

To me different clients to do different things and it is kind of unexpected to see history that you don't see before (so for example bash will not do this, after conifguring it to dump history on each line as described [here](https://github.com/ClickHouse/ClickHouse/pull/13086#issuecomment-665841801))

> So, when you work with many clickhouse-clients, they will see history of each other.

I suggested to make it under option, but as you said it is [too complex](https://github.com/ClickHouse/ClickHouse/pull/13086#issuecomment-665851819) (and I agree, since then this option should be covered)

alexey-milovidov commented:
Ok, reasonable.

azat commented:
@alexey-milovidov can you merge then?
